src/corefx/System.Globalization.Native/pal_collation.c

```diff
@@ -385,38 +377,40 @@ void GlobalizationNative_CloseSortHandle(SortHandle* pSortHandle)
         }
     }
 
-    pthread_mutex_destroy(&pSortHandle->collatorsLockObject);
-
     free(pSortHandle);
 }
 
 const UCollator* GetCollatorFromSortHandle(SortHandle* pSortHandle, int32_t options, UErrorCode* pErr)
 {
-    UCollator* pCollator;
     if (options == 0)
     {
-        pCollator = pSortHandle->collatorsPerOption[0];
+        return pSortHandle->collatorsPerOption[0];
     }
     else
     {
-        int lockResult = pthread_mutex_lock(&pSortHandle->collatorsLockObject);
-        if (lockResult != 0)
+        options &= CompareOptionsMask;
+        UCollator* pCollator = pSortHandle->collatorsPerOption[options];
+        if (pCollator != NULL)
         {
-            assert(FALSE && "Unexpected pthread_mutex_lock return value.");
+            return pCollator;
         }
 
-        options &= CompareOptionsMask;
-        pCollator = pSortHandle->collatorsPerOption[options];
-        if (pCollator == NULL)
+        pCollator = CloneCollatorWithOptions(pSortHandle->collatorsPerOption[0], options, pErr);
+
+#if HAVE_WORKING_STDATOMIC
+        UCollator* pNull = NULL;
+        if (!atomic_compare_exchange_strong((_Atomic(UCollator*)*)&pSortHandle->collatorsPerOption[options], &pNull, pCollator))
```

Comment:
so this is not working on ARM at all? I know you have tried. I am just wondering if there could be a way making it work?